plugins/tedge_apt_plugin/src/main.rs

```diff
@@ -240,23 +251,73 @@ fn main() {
+    // derived from the exec name with which it was called (i.E. a softlink name the original filename).
+    // NOTE: In case module type can not be derived, the plugin shall pani. Thats why unwrap() is used intentionally.
+    let args : Vec<String> = std::env::args().collect();
+    let module_type = std::path::Path::new(& args[0]).file_name().unwrap().to_os_string().into_string().unwrap();
+    let exit_code;
+
+    match run(apt.operation, &module_type) {
         Ok(status) if status.success() => {
-            std::process::exit(0);
+            exit_code = 0;
         }
 
         Ok(status) => {
             if status.code().is_some() {
-                std::process::exit(2);
+                exit_code= 2;
             } else {
                 eprintln!("Interrupted by a signal!");
-                std::process::exit(4);
+                exit_code = 4;
             }
         }
 
         Err(err) => {
             eprintln!("ERROR: {}", err);
-            std::process::exit(5);
+            exit_code = 5;
         }
     }
+
+    if module_type == "tedge" {
+
+        let mut publish_succeeded = false;
+
+        let mqtt_port = 1883; // TODO: read from tedge-config
+
+        // loop until message has arrived at the broker
+        while !publish_succeeded {
+
+            // Using here runmqtt directly instead of tedge's mqtt_channel class, to have access
+            // to runmqtt's eventloop for feedback (PubComp). That is to wait with plugin's process-exit until
+            // published message has really arrived at the broker.
+            let mut options = rumqttc::MqttOptions::new(String::from("sm plugin-apt"), String::from("localhost"), mqtt_port);
+            options.set_clean_session(true);
+            let (mut client, mut connection) = rumqttc::Client::new(options, DEFAULT_QUEUE_CAPACITY);
+            _ = client.publish(String::from("tedge/plugins/software/tedge"), rumqttc::QoS::ExactlyOnce, true, exit_code.to_string());
+
+            for event in connection.iter() {
+                match event {
+
+                    Ok(Event::Incoming(Packet::PubComp(_))) => {
+                      publish_succeeded = true;
+                      break;
+                    }
+
+                    // TODO: Crosscheck that all fail cases are covered!
+                    //       Check for a more safe way to really assure not to run in endless loop?
+                    Ok(Event::Outgoing(Outgoing::Disconnect))
+                    | Ok(Event::Incoming(Incoming::Disconnect))
+                    | Err(_) => {
+                      // Sending pub to broker failed. Wait some time to not overload the CPU and loopback network with MQTT pub retries.
+                      std::thread::sleep(std::time::Duration::from_secs(5));
+                      break;
+                    }
+                    _ => { }
+                }
+            }
+        }
+    }
```

Comment:
There is now a `close` method in `mqttt_channel` which I believe can be used instead.

Reply:
I do think we have to change the approach here. I know that I propose to use MQTT to communicate the outcome to the agent. But this was a bad idea. And this code makes this obvious: there are too many reliability issues while only the agent is expecting this result - there is no benefits to publish it over MQTT. Why not create a file?

Reply:
> This loop is doing more, compared to `mqtt_chanel::Connection::close()`. The latter doesn't re-published the messages on network failure, as done here.

This is true, but we wouldn't have to if as you stated below, just make it a file.




> I do think we have to change the approach here. I know that I propose to use MQTT to communicate the outcome to the agent. But this was a bad idea. And this code makes this obvious: there are too many reliability issues while only the agent is expecting this result - there is no benefits to publish it over MQTT. Why not create a file?

I also think that this adds too much of a dependency and is a little too much for just passing an exit code.
I would really reconsider an option to not restart on upgrade and maybe handle it ourselves instead, then none of that would be needed.

Reply:
> [...] there is no benefits to publish it over MQTT. Why not create a file?

@didier-wenzek , a benefit is that with the MQTT subsciption the SM Agent is by design notified as soon as the result code is available. For the file-based approach some complexity might move into the SM agent, e.g. due to need for polling for the file or use of inotify(). Anyway I would be ok to go back a file-based approach.

However we need to take a decision now, and we must not question it the day after tomorrow again. For that reason I like to address some quesions to @makr11st : 
   - Could you please check what would it mean for the SM Agent to replace MQTT with a file on the disc to transfer the final result code from the tedge-pugin to the SM Agent?
   - Do you see any risc or any challenges we shall discuss before deciding to go that way?
   - What do you think about the effort for the SM Agent implementation? Do you see a significant differences compared to the effort for the MQTT way?

@makr11st , thanks in advance for commenting points above.

Reply:
Indeed that's true, and that simplifies the Agent a lot. To summarize the flow of self-update:

1. agent executes plugin
2. plugin downloads all packages that have some download URL given
3. plugin calls `tedge disconnect` for all bridges
(i.E. old tedge CLI stops old agent and mappers)
4. plugin proceeds insall/remove actions for all packages
5. plugin creates a file containing final exit-code of installation procedure
(filename and path to be aligned)
6. plugin calls `tedge connect` for all bridges that were connected before
(i.E. new tedge CLI starts new agent and mappers)
7. agent starts and reads final exit-code from file.
In parallel, plugin exits.

@makr11st , let go with that filebased approach. I will adapt the spec accordingly.

Reply:
> * Could you please check what would it mean for the SM Agent to replace MQTT with a file on the disc to transfer the final result code from the tedge-pugin to the SM Agent?
> * Do you see any risk or any challenges we shall discuss before deciding to go that way?

This will be much simpler on the agent side.
And the new flow will simplify the recovery as well.